godot-core/src/builtin/basis.rs

```diff
@@ -0,0 +1,980 @@
+        // Godot output:
+        // [X: (1, 0, 0), Y: (0, 1, 0), Z: (0, 0, 1)]
+        // Where X,Y,Z are the columns
+        let [Vector3 {
+            x: x1,
+            y: x2,
+            z: x3,
+        }, Vector3 {
+            x: y1,
+            y: y2,
+            z: y3,
+        }, Vector3 {
+            x: z1,
+            y: z2,
+            z: z3,
+        }] = self.to_cols();
+
+        f.write_fmt(format_args!(
+            "[X: ({x1}, {x2}, {x3}), Y: ({y1}, {y2}, {y3}), Z: ({z1}, {z2}, {z3})]"
+        ))
+    }
+}
+
+impl From<Quaternion> for Basis {
+    fn from(quat: Quaternion) -> Self {
+        Basis::from_quat(quat)
+    }
+}
+
+impl GlamConv for Basis {
+    type Glam = glam::f32::Mat3;
+}
+
+impl GlamType for glam::f32::Mat3 {
+    type Mapped = Basis;
+
+    fn to_front(&self) -> Self::Mapped {
+        Basis::from_rows_array(&self.to_cols_array()).transposed()
+    }
+
+    fn from_front(mapped: &Self::Mapped) -> Self {
+        Self::from_cols_array(&mapped.to_rows_array()).transpose()
+    }
+}
+
+impl GlamType for glam::f32::Mat3A {
+    type Mapped = Basis;
+
+    fn to_front(&self) -> Self::Mapped {
+        Basis::from_rows_array(&self.to_cols_array()).transposed()
+    }
+
+    fn from_front(mapped: &Self::Mapped) -> Self {
+        Self::from_cols_array(&mapped.to_rows_array()).transpose()
+    }
+}
+
+impl Index<usize> for Basis {
+    type Output = Vector3;
+
+    /// Indexes into the basis by rows.
+    #[inline]
+    fn index(&self, index: usize) -> &Self::Output {
+        &self.rows[index]
+    }
+}
+
+impl IndexMut<usize> for Basis {
+    /// Indexes into the basis by rows.
+    #[inline]
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        &mut self.rows[index]
+    }
+}
+
+impl Default for Basis {
+    fn default() -> Self {
+        Self::IDENTITY
+    }
+}
+
+impl Mul for Basis {
+    type Output = Self;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        self.glam2(&rhs, |a, b| a * b)
+    }
+}
+
+impl Mul<f32> for Basis {
+    type Output = Self;
+
+    fn mul(self, rhs: f32) -> Self::Output {
+        (self.to_glam() * rhs).to_front()
+    }
+}
+
+impl MulAssign<f32> for Basis {
+    fn mul_assign(&mut self, rhs: f32) {
+        self[0] *= rhs;
+        self[1] *= rhs;
+        self[2] *= rhs;
+    }
+}
+
+impl Mul<Vector3> for Basis {
+    type Output = Vector3;
+
+    fn mul(self, rhs: Vector3) -> Self::Output {
+        self.glam2(&rhs, |a, b| a * b)
+    }
+}
+
+impl GodotFfi for Basis {
+    ffi_methods! { type sys::GDExtensionTypePtr = *mut Self; .. }
+}
+
+/// The ordering used to interpret a set of euler angles as extrinsic
+/// rotations.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(C)]
+pub enum EulerOrder {
+    XYZ = 0,
+    XZY = 1,
+    YXZ = 2,
+    YZX = 3,
+    ZXY = 4,
+    ZYX = 5,
+}
+
+impl From<EulerOrder> for i64 {
+    fn from(order: EulerOrder) -> Self {
+        order as i64
+    }
+}
```

Comment:
This is only needed for the implementation, as far as I can see. As such, we don't need a `From` conversion and can just use `as` internally (or an inherent, private `to_sys() -> i64` method).

Reply:
It's actually for an itest because of a limitation of the `assert_eq_approx` macro.

Reply:
The problem is that this leaks into the public API...

I'm not sure if you can apply `#[doc(hidden)]` to trait impls, could you try that? And if not, maybe add a TODO?

Reply:
It appears you can